src/core/lib/security/credentials/tls/grpc_tls_certificate_provider.cc

```diff
@@ -198,13 +198,38 @@ void FileWatcherCertificateProvider::ForceUpdate() {
       root_certificate_ = "";
     }
   }
+  absl::StatusOr<bool> keyCertMatch = false;
+  //check whether the present credentials are valid.
+  if (!pem_key_cert_pairs_.empty()) {
+    for (int i = 0; i < pem_key_cert_pairs_.size(); i++) {
+      keyCertMatch =
+          PrivateKeyAndCertificateMatch(pem_key_cert_pairs_.at(i).private_key(),
+                                        pem_key_cert_pairs_.at(i).cert_chain());
+      if (!(keyCertMatch.ok() && *keyCertMatch)) {
+        break;
+      }
+    }
+  }
   const bool identity_cert_changed =
       (!pem_key_cert_pairs.has_value() && !pem_key_cert_pairs_.empty()) ||
```

Comment:
`identity_cert_changed` will be `true` if the new pair is empty and the old one is not, which is undesired. We wouldn't want to replace an old pair with an empty one.

cc: @ZhenLian

Reply:
I think that's to enable deletion of credentials, or am I missing something here?
PS: I didn't change the conditions for identity_cert_changed, they've always been like that

Reply:
Yeah, they were like that but they don't go with our current design approach, I believe. Best talk to @ZhenLian .

Reply:
That's an interesting topic. At first(before the always-failing handshaker went out), I thought we better do the check of empty creds, since that are blatantly misuse of our API and we'd better avoid our program to crash.
With the always-failing handshaker, if that happens, our program won't crash - the server would only drop any connections it takes. That makes me wonder if there could be a certain period of time when users want to clear their creds and make use of the always-failing handshaker. Though I would say that case might be rare and probably might never exist.
But the point is to not "help" users to make the decision. If they at some point clear their cred files, then we shall just clear our data as well. Users might see this as unexpected, but that's because they clear the files in the first place - which is already a mis-behavior.
The same logic applies to the `DataWatcherCertificateProvider` as well.